### ISSUE TITLE
feat(`cheatcodes`): add `expect*Call` cheatcodes

### ISSUE THREAD

**Issue by grandizzy** [foundry-rs/foundry#9310]
### Component

Forge

### Describe the feature you would like

Follow up of https://github.com/foundry-rs/foundry/issues/4072 for adding remaining `expect*Call` cheatcodes

- [ ] `expect*Call` cheatcodes: similar to expectEmit, these let us specify the expected call with its callee and calldata. We use these less frequently, but I think they can be useful even for fuzzing tests.

  - [ ] `expectRegularCall`: Expect a call using the `CALL` opcode and specified parameters
  - [ ] `expectStaticCall`: Expect a call using the `STATICCALL` opcode and specified parameters
  - [ ] `expectDelegateCall`: Expect a call using the `DELEGATECALL` opcode and specified parameters
  - [ ] `expectNoCall`: Expect no calls to be made after this cheat code
  - [ ] `expectCreate`: Expects the deployment of the specified bytecode by the specified address using the `CREATE` opcode
  - [ ] `expectCreate2`: Expects the deployment of the specified bytecode by the specified address using the `CREATE2` opcode

**Comment by Utilitycoder** [comment]
I would love to handle this issue. I will probably need the files that will be affected and where to draw inspiration from in the codebase.

**Comment by mimisavage** [comment]
Can I work on this, please?

**Comment by Nityam573** [comment]
May I take care of this?

**Comment by anukul** [comment]
@grandizzy hi! i'm interested in working on this :)

**Comment by grandizzy** [comment]
@anukul awesome, thanks! I assigned it to you

**Comment by anukul** [comment]
From the existing `expectCall` overloads, which is the ideal set of parameters to use for each  `expect*Call` cheatcode? 🤔 

https://github.com/foundry-rs/foundry/blob/dabacecdc14d074a108c18f97d1e1f63ade37a37/crates/cheatcodes/spec/src/vm.rs#L930-L962

There is this option:

```solidity
function expectRegularCall(address callee, uint256 msgValue, uint64 gas, bytes calldata data) external;
function expectRegularCall(address callee, uint256 msgValue, uint64 gas, bytes calldata data, uint64 count) external;
```

or the possibility of another function, something like:

```solidity
struct CallType {
  Regular,
  Static,
  Delegate
}

struct ExpectCallOptions {
    CallType kind;

    // optional gas fields (0 = don't check)
    uint64 exactGas;
    uint64 minGas;
    // optional count (0 = expect at least once)
    uint64 count;
    
    // optional msg.value
    uint256 msgValue;
    // flag for ambiguous zero value fields
    bool checkMsgValue;
}

function expectCall(
    address callee,
    bytes calldata data,
    ExpectCallOptions calldata options
) external;
```

**Comment by anukul** [comment]
Somewhat related - https://github.com/foundry-rs/foundry/issues/8845, to make `data` optional

**Comment by Abhishekkochar** [comment]
Hi @anukul, I can take this over if you occupied.

**Comment by anukul** [comment]
hi @Abhishekkochar, thanks for flagging!

i think it could be valuable to discuss the `expect*Call` cheatcodes before implementation - perhaps a way to reduce the combinatorial explosion in number of overloads from the different optional parameters.

i can implement the rest of the functions in the coming week :)

**Comment by Abhishekkochar** [comment]
> hi [@Abhishekkochar](https://github.com/Abhishekkochar), thanks for flagging!
> 
> i think it could be valuable to discuss the `expect*Call` cheatcodes before implementation - perhaps a way to reduce the combinatorial explosion in number of overloads from the different optional parameters.
> 
> i can implement the rest of the functions in the coming week :)

If you waiting on team, they may not get back to you in time. And, for `expect*Call` cheatcode, it has to be follow the similar pattern as `ExpectEmit`. I can have this implemented in a week or so, if you are occupied.

**Comment by anukul** [comment]
@Abhishekkochar I'm not sure if that was a question. If so, please refer to my last comment.

**Comment by grandizzy** [comment]
> 
> or the possibility of another function, something like:
> 
> enum CallType {
>   Regular,
>   Static,
>   Delegate
> }
> 
> struct ExpectCallOptions {
>     CallType kind;
> 
>     // optional gas fields (0 = don't check)
>     uint64 exactGas;
>     uint64 minGas;
>     // optional count (0 = expect at least once)
>     uint64 count;
>     
>     // optional msg.value
>     uint256 msgValue;
>     // flag for ambiguous zero value fields
>     // or alternatively, an OptionalUint256 { bool hasValue, uint256 value} 
>     bool checkMsgValue;
> }
> 

@anukul since fields in `ExpectCallOptions` (but call type) are optional I'd suggest to have similar as https://github.com/foundry-rs/forge-std/blob/bf909b22fa55e244796dfa920c9639fdffa1c545/src/Vm.sol#L2125-L2142 like

```Solidity
    function expectCall(CallKind callKind, address callee, bytes calldata data) external;
    function expectCall(CallKind callKind, address callee, bytes calldata data, uint64 count) external;
    function expectCall(CallKind callKind, address callee, uint256 msgValue, bytes calldata data) external;
    function expectCall(CallKind callKind, address callee, uint256 msgValue, bytes calldata data, uint64 count) external;
    function expectCall(CallKind callKind, address callee, uint256 msgValue, uint64 gas, bytes calldata data) external;
    function expectCall(CallKind callKind, address callee, uint256 msgValue, uint64 gas, bytes calldata data, uint64 count) external;
```
and at some point we could drop the existing `expectCall`s and point to using the new one with passing regular call. @zerosnacks pls chime in

**Comment by zerosnacks** [comment]
I agree with the proposed functions, using function overrides like this tends to be more convenient for end users

```solidity
    function expectCall(CallKind callKind, address callee, bytes calldata data) external;
    function expectCall(CallKind callKind, address callee, bytes calldata data, uint64 count) external;
    function expectCall(CallKind callKind, address callee, uint256 msgValue, bytes calldata data) external;
    function expectCall(CallKind callKind, address callee, uint256 msgValue, bytes calldata data, uint64 count) external;
    function expectCall(CallKind callKind, address callee, uint256 msgValue, uint64 gas, bytes calldata data) external;
    function expectCall(CallKind callKind, address callee, uint256 msgValue, uint64 gas, bytes calldata data, uint64 count) external;
```
